lib/executor/variable_arrival_rate.go

```diff
@@ -329,6 +328,28 @@ func (varr VariableArrivalRate) Run(ctx context.Context, out chan<- stats.Sample
 		return activeVU
 	}
 
+	remainingUnplannedVUs := maxVUs - preAllocatedVUs
+	makeUnplannedVUCh := make(chan struct{})
+	defer close(makeUnplannedVUCh)
+	go func() {
+		defer close(returnedVUs)
+		defer func() {
+			// this is done here as to not have an unplannedVU in the middle of initialization when
+			// starting to return activeVUs
+			for i := uint64(0); i < atomic.LoadUint64(&activeVUsCount); i++ {
+				<-activeVUs
+			}
+		}()
+		for range makeUnplannedVUCh {
+			initVU, err := varr.executionState.GetUnplannedVU(maxDurationCtx, varr.logger)
+			if err != nil {
+				// TODO figure out how to return it to the Run goroutine
+				varr.logger.WithError(err).Error("Error while allocating unplanned VU")
```

Comment:
This will need fixing, but I guess I would refactor most of this into a helper struct/type thingy that does all the getting of Vus and returning them and waiting and so on ... in a more ... readable matter :(